output/shared/chunkfactory.go

```diff
@@ -0,0 +1,126 @@
+package shared
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/relex/gotils/logger"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/util"
+)
+
+type InitCompessorFunc func(log logger.Logger, w io.Writer) io.WriteCloser
+
+type chunkEncoder interface {
+	EncodeChunk(chunk *BasicChunk) ([]byte, error)
+}
+
+type BasicChunkFactory struct {
+	log               logger.Logger
+	reusedChunkBuffer *bytes.Buffer // since chunks are created consecutively, it's possible for them to share the data buffer
+	initCompessor     InitCompessorFunc
+	idGenerator       *chunkIDGenerator
+	chunkEncoder      chunkEncoder
+}
+
+type BasicChunk struct {
+	id           string
+	numRecords   int
+	numBytes     int
+	compressor   io.WriteCloser // could be something like a gzip.Writer or nil to disable compression
+	reusedBuffer *bytes.Buffer  // an actual buffer that compressor writes to
+	encoder      chunkEncoder   // a function to finalize/encode the entire chunk after it's been assembled
+}
+
+func NewChunkFactory(log logger.Logger, idSuffix string, msgBufCapacity int, initCompessor InitCompessorFunc, encoder chunkEncoder) *BasicChunkFactory {
+	return &BasicChunkFactory{
+		log:               log,
+		reusedChunkBuffer: bytes.NewBuffer(make([]byte, 0, msgBufCapacity)),
+		initCompessor:     initCompessor,
+		idGenerator:       newChunkIDGenerator(idSuffix),
+		chunkEncoder:      encoder,
+	}
+}
+
+func (factory *BasicChunkFactory) NewChunk() *BasicChunk {
+	chunk := &BasicChunk{
+		id:           factory.idGenerator.Generate(),
+		numRecords:   0,
+		numBytes:     0,
+		compressor:   nil,
+		reusedBuffer: factory.reusedChunkBuffer,
+		encoder:      factory.chunkEncoder,
+	}
+
+	if factory.initCompessor != nil {
+		chunk.compressor = factory.initCompessor(factory.log, factory.reusedChunkBuffer)
+	}
+
+	return chunk
+}
+
+// Write appends new log to log chunk
+func (chunk *BasicChunk) Write(data []byte) error {
```

Comment:
The `data` could be better named to represent serialized data of a single log record, or the type can be `LogStream` (alias to []byte).